lib/elixir/test/elixir/module/types/expr_test.exs

```diff
@@ -263,6 +263,87 @@ defmodule Module.Types.ExprTest do
              )
   end
 
+  test "map update" do
+    assert quoted_expr(
+             (
+               map = %{foo: :a}
+               %{map | foo: :b}
+             )
+           ) ==
+             {:ok, {:map, [{:required, {:atom, :foo}, {:atom, :b}}]}}
+
+    assert quoted_expr([map], %{map | foo: :b}) ==
+             {:ok,
+              {:map, [{:required, {:atom, :foo}, {:atom, :b}}, {:optional, :dynamic, :dynamic}]}}
+
+    assert {:error,
+            {:unable_unify,
+             {{:map, [{:required, {:atom, :bar}, :dynamic}, {:optional, :dynamic, :dynamic}]},
```

Comment:
The downside with using `dynamic()` for the values is that it shows up in the error messages which could be slightly confusing.